File: Source/TagDeleter.cs
namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that deletes specified tags from files.
/// </summary>
public class TagDeleter : TagBase
{
    /// <summary>
    /// Executes the tag deletion process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path and the second argument is the file containing tags to remove.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        var removeWordsFileName = args.Count >= 2 ? args[1] : "";
        var removeWords = !string.IsNullOrEmpty(removeWordsFileName) ? File.ReadAllLines(removeWordsFileName).ToList() : new List<string>();

        foreach (var fileName in fileNames)
            ProcessFile(fileName, removeWords);
    }

    /// <summary>
    /// Processes the specified file to remove specified tags.
    /// </summary>
    /// <param name="fileName">The name of the file to process.</param>
    /// <param name="removeWords">A list of tags to remove from the file.</param>
    static void ProcessFile(string fileName, List<string> removeWords)
    {
        var lines = File.ReadAllLines(fileName).ToList();

        for (var i = 0; i < lines.Count; i++)
        {
            var line = lines[i];
            var newLine = new List<string>();

            var items = TagHelper.SplitIntoTags(line);
            for (var j = 0; j < items.Count; j++)
            {
                var item = items[j];
                item = item.Trim();
                if (!removeWords.Contains(item) && !string.IsNullOrEmpty(item))
                {
                    newLine.Add(item);
                }
            }
            newLine = newLine.ToList();
            line = string.Join(", ", newLine);
            lines[i] = line;
        }

        File.WriteAllLines(fileName, lines);
    }
}

File: tagbag/tag_deleter.py
from .tag_base import TagBase
from .tag_helper import TagHelper


from pathlib import Path
from typing import List, Sequence, Set


class TagDeleter(TagBase):
    """
    Delete specified tags (listed one per line in a separate file) from each line.
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        args[1] optional: text file containing tags to remove, one per line
        """
        dir_path = args[0]
        remove_words_file = args[1] if len(args) >= 2 else ""
        remove_words: List[str] = []
        if remove_words_file:
            remove_words = Path(remove_words_file).read_text(encoding="utf-8").splitlines()

        file_names = TagBase.get_tag_files(dir_path)
        remove_set = {w.strip() for w in remove_words if w.strip()}
        for name in file_names:
            TagDeleter._process_file(Path(name), remove_set)

    @staticmethod
    def _process_file(file_path: Path, remove_words: Set[str]) -> None:
        lines = file_path.read_text(encoding="utf-8").splitlines()
        out_lines: List[str] = []

        for line in lines:
            kept: List[str] = []
            for item in TagHelper.split_into_tags(line):
                tag = item.strip()
                if tag and tag not in remove_words:
                    kept.append(tag)
            out_lines.append(", ".join(kept))

        file_path.write_text("\n".join(out_lines), encoding="utf-8")